hearth/lib/hearth/connection_pool.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+module Hearth
+  # @api private
+  class ConnectionPool
+    @pools_mutex = Mutex.new
+    @pools = {}
+
+    class << self
+      # @return [ConnectionPool]
+      def for(config = {})
```

Comment:
Why are `self.for` and `session_for` not combined as just `lease`?

Reply:
I don't think we would want to create an instance of ConnectionPool because the pools need to maintain state across multiple clients potentially. So a lease method would need to be on the class, but then it would need to understand a particular instance. What would that look like?

Reply:
I was suggesting a `self.lease` that would do somethin glike:
```
# skips mutex handling for brevity
def self.lease(config, endpoint, &block)
  pool = @pools[config] ||= new
  pool.session_for(endpoint, &block)
end
```

but with the optimization suggested below (moving getting the pool into the http client initialization, this doesn't make sense, so nvm)

Reply:
Moving pool into http client initialization would mean that it doesn't work for request level options